contributors/design-proposals/node/node-usernamespace-remapping.md

```diff
@@ -0,0 +1,210 @@
+- CRI-O: WIP https://github.com/nalind/cri-o/tree/idmapping
+
+### Implementation Roadmap
+#### Phase 1: Support in Kubelet, Alpha, [Target: Kubernetes v1.11]
+- Add feature gate `NodeUserNamespace`, disabled by default
+- Add new CRI API, `RuntimeConfigInfo()`
+- Add logic in Kubelet to handle pod creation which includes parsing RuntimeConfigInfo response and changing file-permissions in /var/lib/kubelet with learned userns mapping.
+- Add changes in dockershim to implement RuntimeConfigInfo() for docker runtime
+- Add changes in CRI-O to implement userns support and RuntimeConfigInfo() support
+- Unit test cases
+
+#### Phase 2: Beta Support [Target: Kubernetes v1.12]
+- PSP integration
+- To grow ExperimentalHostUserNamespaceDefaulting from experimental feature gate to a Kubelet flag
+- API changes to allow pod able to request HostUserNamespace in pod spec
+- e2e tests
+
+### References
+- Default host user namespace via experimental flag
+  - https://github.com/kubernetes/kubernetes/pull/31169
+- Enable userns support for containers launched by kubelet
+  - https://github.com/kubernetes/features/issues/127
+- Track Linux User Namespaces in the Pod Security Policy
+  - https://github.com/kubernetes/kubernetes/issues/59152
+- Add support for experimental-userns-remap-root-uid and experimental-userns-remap-root-gid options to match the remapping used by the container runtime.
+  - https://github.com/kubernetes/kubernetes/pull/55707
+- rkt User Namespaces Background
+  - https://coreos.com/rkt/docs/latest/devel/user-namespaces.html
+
+## Future Work
+- Named User namespaces: There is no runtime today which supports creating containers with a named user namespace. For example here is the discussion related to this support in Docker https://github.com/moby/moby/issues/28593 and support is being created for CRI-O https://github.com/nalind/cri-o/tree/idmapping.
+- Once user-namespace feature in the runtimes has evolved to support container’s request for a specific user-namespace mapping(UID and GID range) by name(range identifier), Kubernetes might leverage this to support Namespace level isolation by allocating a different runtime user-namespace mapping (UID and GID range) to each Kubernetes namespace.
+
+## Risks and Mitigations
+The main risk with this change stems from the fact that processes in Pods will run with different “real” uids than they used to, while expecting the original uids to make operations on the Nodes or consistently access shared persistent storage.
+- This can be mitigated by turning the feature on gradually, per-Pod or per Kubernetes namespace.
+- For the Kubernetes' cluster Pods (that provide the Kubernetes functionality), testing of their behaviour and ability to run in user namespaced setups is crucial.
+
+## Graduation Criteria
+- PSP integration
+- API changes to allow pod able to request host user namespace using for example, `HostUserNamespace: True`, in pod spec
```

Comment:
Is there a use case for this outside of the items mentioned in defaulting (priv, other host ns, host volumes, etc)?

Reply:
It is pretty much to be implicit about using the host / `NODE` user namespace, without having rely on the defaulting behaviour. For a start, the reasons might be those that you've listed.

Long term, we will probably want the default to be for every Pod to get unique per-Pod user namespace, unless the Pod Spec requests some specific user namespace mapping, or unless the Pod Spec requests to run with host / `NODE` user namespace, for whatever reason -- just for root to be a host root.

Reply:
I think handling what users mean by asking for privileged is hard to figure out automatically. It could imply give me all CAPs in a non host user namespace or give me all CAPs in the host user namespace.